models/sessions/scratch/bigquery/snowplow_web_sessions_this_run.sql

```diff
@@ -0,0 +1,303 @@
+    case when b.last_page_url is null then coalesce(b.last_page_title, a.first_page_title) else b.last_page_title end as last_page_title,
+    case when b.last_page_url is null then coalesce(b.last_page_url, a.first_page_url) else b.last_page_url end as last_page_url,
+    case when b.last_page_url is null then coalesce(b.last_page_urlscheme, a.first_page_urlscheme) else b.last_page_urlscheme end as last_page_urlscheme,
+    case when b.last_page_url is null then coalesce(b.last_page_urlhost, a.first_page_urlhost) else b.last_page_urlhost end as last_page_urlhost,
+    case when b.last_page_url is null then coalesce(b.last_page_urlpath, a.first_page_urlpath) else b.last_page_urlpath end as last_page_urlpath,
+    case when b.last_page_url is null then coalesce(b.last_page_urlquery, a.first_page_urlquery) else b.last_page_urlquery end as last_page_urlquery,
+    case when b.last_page_url is null then coalesce(b.last_page_urlfragment, a.first_page_urlfragment) else b.last_page_urlfragment end as last_page_urlfragment,
+
+    -- referrer fields
+    a.referrer,
+    a.refr_urlscheme,
+    a.refr_urlhost,
+    a.refr_urlpath,
+    a.refr_urlquery,
+    a.refr_urlfragment,
+    a.refr_medium,
+    a.refr_source,
+    a.refr_term,
+
+    -- marketing fields
+    a.mkt_medium,
+    a.mkt_source,
+    a.mkt_term,
+    a.mkt_content,
+    a.mkt_campaign,
+    a.mkt_clickid,
+    a.mkt_network,
+
+    -- geo fields
+    a.geo_country,
+    a.geo_region,
+    a.geo_region_name,
+    a.geo_city,
+    a.geo_zipcode,
+    a.geo_latitude,
+    a.geo_longitude,
+    a.geo_timezone,
+
+    -- ip address
+    a.user_ipaddress,
+
+    -- user agent
+    a.useragent,
+
+    a.br_renderengine,
+    a.br_lang,
+
+    a.os_timezone,
+
+    -- optional fields, only populated if enabled.
+    -- iab enrichment fields
+    a.category,
+    a.primary_impact,
+    a.reason,
+    a.spider_or_robot,
+
+    -- ua parser enrichment fields
+    a.useragent_family,
+    a.useragent_major,
+    a.useragent_minor,
+    a.useragent_patch,
+    a.useragent_version,
+    a.os_family,
+    a.os_major,
+    a.os_minor,
+    a.os_patch,
+    a.os_patch_minor,
+    a.os_version,
+    a.device_family,
+
+    -- yauaa enrichment fields
+    a.device_class,
+    a.agent_class,
+    a.agent_name,
+    a.agent_name_version,
+    a.agent_name_version_major,
+    a.agent_version,
+    a.agent_version_major,
+    a.device_brand,
+    a.device_name,
+    a.device_version,
+    a.layout_engine_class,
+    a.layout_engine_name,
+    a.layout_engine_name_version,
+    a.layout_engine_name_version_major,
+    a.layout_engine_version,
+    a.layout_engine_version_major,
+    a.operating_system_class,
+    a.operating_system_name,
+    a.operating_system_name_version,
+    a.operating_system_version
+from
+    my_firsts a
+left join
+    my_lasts b on a.domain_sessionid = b.domain_sessionid and b.page_event_in_session_index = 1
```

Comment:
We're currently not filtering at all within the ctes for `my_firsts` and `my_lasts`, couldn't we just do a self join to grab the first and last and save ourselves a cte?

Reply:
We filter first to both `page_view` and `page_ping` events as where `my_lasts` is only filtered to `page_views` (after discussions with Jordan and to align to current processing). There is maybe a clever way to do it but it feels like being _too_ smart about it, and this helps keep the queries structuraly similar across warehouses at least?